embassy-stm32f4/src/exti.rs

```diff
@@ -0,0 +1,128 @@
+//! Async low power Serial.
+//!
+//! The peripheral is autmatically enabled and disabled as required to save power.
+//! Lowest power consumption can only be guaranteed if the send receive futures
+//! are dropped correctly (e.g. not using `mem::forget()`).
+
+use core::future::Future;
+use core::mem;
+use core::pin::Pin;
+
+use embassy::gpio::{WaitForFallingEdge, WaitForRisingEdge};
+use embassy::interrupt::OwnedInterrupt;
+use embassy::util::Signal;
+
+use crate::hal::gpio::{Edge, ExtiPin as HalExtiPin};
+use crate::hal::prelude::*;
+use crate::hal::syscfg::SysCfg;
+use crate::pac::EXTI;
+
+use crate::interrupt;
+
+static mut exti: mem::MaybeUninit<EXTI> = mem::MaybeUninit::<EXTI>::uninit();
+
+
+pub struct ExtiContainer {}
+
+impl ExtiContainer {
+    pub fn new(ex: EXTI) -> Self {
+        unsafe { exti.write(ex); }
+
+        Self{}
+    }
+}
+
+
+/// Interface to the Serial peripheral
+pub struct ExtiPin<T: HalExtiPin, I: OwnedInterrupt> {
+    pin: T,
+    signal: Signal<()>,
+    interrupt: I,
+}
+
+/*
+    Irq	Handler	Description
+    EXTI0_IRQn	EXTI0_IRQHandler	Handler for pins connected to line 0
+    EXTI1_IRQn	EXTI1_IRQHandler	Handler for pins connected to line 1
+    EXTI2_IRQn	EXTI2_IRQHandler	Handler for pins connected to line 2
+    EXTI3_IRQn	EXTI3_IRQHandler	Handler for pins connected to line 3
+    EXTI4_IRQn	EXTI4_IRQHandler	Handler for pins connected to line 4
+    EXTI9_5_IRQn	EXTI9_5_IRQHandler	Handler for pins connected to line 5 to 9
+    EXTI15_10_IRQn	EXTI15_10_IRQHandler	Handler for pins connected to line 10 to 15
+*/
+
+impl<T: HalExtiPin, I: OwnedInterrupt> ExtiPin<T, I> {
+    pub fn new(mut pin: T, container: &ExtiContainer, syscfg: &mut SysCfg, interrupt: I) -> Self {
+        pin.make_interrupt_source(syscfg);
+
+        ExtiPin {
+            pin: pin,
+            signal: Signal::new(),
+            interrupt: interrupt,
+        }
+    }
+
+    unsafe fn on_interrupt(&self) {
+        self.interrupt.disable();
+        self.signal.signal(());
+    }
+}
+
+impl<T: HalExtiPin + 'static, I: OwnedInterrupt + 'static> WaitForRisingEdge for ExtiPin<T, I> {
+    type Future<'a> = impl Future<Output = ()> + 'a;
+
+    fn wait_for_rising_edge<'a>(self: Pin<&'a mut Self>) -> Self::Future<'a> {
+        let s = unsafe { self.get_unchecked_mut() };
+
+        async move {
+            s.signal.reset();
+            s.pin.clear_interrupt_pending_bit();
+            s.interrupt.set_handler(
+                |ptr| unsafe {
+                    let this = &*(ptr as *const () as *const Self);
+                    this.on_interrupt();
+                },
+                s as *const _ as *mut _,
+            );
+            s.interrupt.unpend();
+            s.interrupt.enable();
+            unsafe {
+                s.pin.trigger_on_edge(&mut *exti.as_mut_ptr(), Edge::RISING);
```

Comment:
You could transmute the EXTI instance "from thin air" because it's zero-sized. This way there's no need for a `static mut`. It's safe for the same reasons reading the `static mut` is safe.

```rust
let exti: EXTI = unsafe { mem::transmute(()) };
```